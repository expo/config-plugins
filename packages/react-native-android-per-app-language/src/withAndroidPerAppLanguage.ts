import {
  ConfigPlugin,
  withAndroidManifest,
  AndroidManifest,
  withDangerousMod,
  AndroidConfig,
} from "expo/config-plugins";
import * as fs from "fs";
import * as path from "path";

type SupportedLanguages = string[];

export const withAndroidPerAppLanguage: ConfigPlugin<{
  supportedLanguages: SupportedLanguages;
}> = (config, { supportedLanguages }) => {
  if (!supportedLanguages || !supportedLanguages.length) {
    return config;
  }

  /**
   * Modify the Android Manifest 'application' to include the localeConfig attribute
   */
  config = withAndroidManifest(config, (config) => {
    config.modResults = addLocaleConfigToManifest(config.modResults);
    return config;
  });

  /**
   * Create `network_security_config.xml` resource file.
   */
  return withDangerousMod(config, [
    "android",
    async (config) => {
      const folder = path.join(
        config.modRequest.platformProjectRoot,
        `app/src/main/res/xml`,
      );
      fs.mkdirSync(folder, { recursive: true });
      fs.writeFileSync(
        path.join(folder, "locales_config.xml"),
        getTemplateFile(supportedLanguages),
        { encoding: "utf8" },
      );
      return config;
    },
  ]);
};

export const addLocaleConfigToManifest = (androidManifest: AndroidManifest) => {
  const application =
    AndroidConfig.Manifest.getMainApplicationOrThrow(androidManifest);

  application.$["android:localeConfig"] = "@xml/locales_config";
  return androidManifest;
};

export const getTemplateFile = (supportedLanguages: SupportedLanguages) => {
  const localesList = supportedLanguages
    .map((lang) => `   <locale android:name="${lang}"/>`)
    .join("\n");

  const content = `<?xml version="1.0" encoding="utf-8"?>
<locale-config xmlns:android="http://schemas.android.com/apk/res/android">
${localesList}
</locale-config>`;
  return content;
};
