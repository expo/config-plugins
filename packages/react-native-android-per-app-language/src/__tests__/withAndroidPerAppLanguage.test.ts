import { getTemplateFile } from "../withAndroidPerAppLanguage";

describe("getTemplateFile", () => {
  it("returns the expected XML content", () => {
    const supportedLanguages = ["en", "fr"];
    const expectedContent = `<?xml version="1.0" encoding="utf-8"?>
<locale-config xmlns:android="http://schemas.android.com/apk/res/android">
   <locale android:name="en"/>
   <locale android:name="fr"/>
</locale-config>`;
    const content = getTemplateFile(supportedLanguages);
    expect(content).toEqual(expectedContent);
  });
});
