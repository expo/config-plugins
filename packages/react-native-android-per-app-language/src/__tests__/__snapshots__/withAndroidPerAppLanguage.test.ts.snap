// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTemplateFile returns only base-config when subdomains is '*' 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
<locale-config xmlns:android="http://schemas.android.com/apk/res/android">
   <locale android:name="en"/>
</locale-config>"
`;
